Improved LDAP and coproprate proxy configurations documentation (#10906)
geosolutions-it/MapStore2: docs/developer-guide/integrations/users/ldap.md

The LDAP storage can be configured in two ways:

* *Synchronized* mode
* *Direct connection* mode (experimental)

### Synchronized mode

In *synchronized mode*, user data (users, groups, roles) is read from LDAP on every login and copied on the internal database.

Any other operation, for example getting user permissions on maps, always uses the internal database.

Synchronized mode is faster for normal use, but data may dis-align when users are removed from the LDAP repository.

In general we suggest to use synchronized mode, since it is the most stable and tested one.

```mermaid
sequenceDiagram
    actor User
    participant MapStore
    participant LDAP
    participant InternalDB

    User->>MapStore: Login with credentials
    MapStore->>LDAP: Check credentials
    LDAP-->>MapStore: Credentials valid
    MapStore->>MapStoreDB: Store user and groups
    MapStore-->>User: Login successful
```

### Direct connection mode (experimental)

In direct connection mode, user data is always read from LDAP, for any operation, so there is no risk of misaligned data.

> Direct connection is still experimental and not tested in all the possible scenarios, but will hopefully become the standard mode in an future, because the approach is simpler and avoids most the synchronized mode defects (e.g. misalignments).

Reviewer: ```suggestion
Direct connection approach is simpler and avoids most the synchronized mode defects (e.g. misalignments).
```

Reviewer: @offtherailz it not yet experimental, we have been using it in production for many years.

* `ldap.roleFilter`: the LDAP filter used to search for role membership of a given user ({0} is the full user DN)
* `ldap.hierachicalGroups`: enables / disables support for nested (hierarchical) groups; when true, a user is assigned groups recursively if its groups belong to other groups
* `ldap.nestedGroupFilter`: the LDAP filter used to search for groups membership of a given group ({0} is the full group DN)
* `ldap.nestedGroupLevels`: the max number of nested groups that are used
* `ldap.searchSubtree`: if true, all the searches are recursive from the relative root path
* `ldap.convertToUpperCase`: if true, all users, groups and roles names are transformed to uppercase in MapStore.

Here a sample `ldap.properties` file:

```properties
ldap.host=localhost
ldap.port=10389
ldap.userDn=
ldap.password=
ldap.root=dc=acme,dc=org

ldap.userBase=ou=people
ldap.groupBase=ou=groups
ldap.roleBase=ou=groups
ldap.userFilter=(uid={0})
ldap.groupFilter=(member={0})
ldap.roleFilter=(member={0})

ldap.hierachicalGroups=false
ldap.nestedGroupFilter=(member={0})
ldap.nestedGroupLevels=3

ldap.searchSubtree=true
ldap.convertToUpperCase=true
```

### Assigning roles

> In the `ldap.properties` you can setup how to assign roles to various users the roles and groups.

Reviewer: ```suggestion
In the `ldap.properties` you can setup how to assign roles to various users and groups.
```

> MapStore has only two roles. `USER` and `ADMIN`. The default `geostore-spring-security-ldap.xml` included in MapStore using the `ldap` profile will look for the `ROLE_ADMIN` in the LDAP sub-tree filtered by the `ldap.roleFilter` property.

Reviewer: ```suggestion
MapStore has only two roles: `USER` and `ADMIN`. The default `geostore-spring-security-ldap.xml`, included in MapStore using the `ldap` profile, will look for the `ROLE_ADMIN` in the LDAP sub-tree filtered by the `ldap.roleFilter` property.
```

If the user belongs to this group, it is assigned the ADMIN role, otherwise the USER role.

> From the recent of `geostore-spring-security-ldap.xml` you can customize the role assignment by editing some special `ldap.properties` file.

Reviewer: ```suggestion
From the `geostore-spring-security-ldap.xml` you can customize the role assignment by editing some special `ldap.properties` file.
```

* `ldap.rolePrefix` is the prefix used for the roles in the LDAP tree. By default it is `ROLE_`.
* `ldap.adminRole` is the role name that assigns the ADMIN role. By default it is `ROLE_ADMIN`.

Here a sample `ldap.properties` extract that will look for the `MAPSTORE_ROLE_ADMIN` role in the LDAP tree to assign the ADMIN role:

```properties
ldap.rolePrefix=MAPSTORE_ROLE_
ldap.adminRole=MAPSTORE_ROLE_ADMIN
```

In previous version this required to edit `geostore-spring-security-ldap.xml` file, by editing `rolePrefix` and adding `authoritiesMapper` property bean to the `GeoStoreLdapAuthoritiesPopulator` bean.

### Assigning groups

> In the `ldap.properties` the properties `ldap.groupBase` and `ldap.groupFilter` are used to search for groups the user belongs to. The `ldap.groupFilter` is used to search for groups the user belongs to. All the groups found are assigned to the user. If using the synchronized mode, the groups are updated in the internal database on every login.

Reviewer: ```suggestion
In the `ldap.properties` the properties `ldap.groupBase` and `ldap.groupFilter` are used to search for groups the user belongs to. The `ldap.groupFilter` is used to search for groups the user belongs to. All the groups found are assigned to the user. Using the synchronized mode, the groups are updated in the internal database on every login.
```

    <!-- -->
```

In the direct mode, all the users and groups are read from LDAP, replacing the internal database of users and groups. In this case, the properties involved are:

* `ldap.userBase`: the root path for searching users. Can be empty.
* `ldap.groupBase`: the root path for searching groups. Can be empty.
* `ldap.memberPattern`: a regular expression to match the user's DN in the group's `member` attribute. By default it is `^uid=([^,]+).*$`. This is used to extract the user's `uid` from the `member` attribute of the group.

Other properties are still to be configured in `geostore-spring-security.xml` file.

## Testing LDAP support

> If you don't have an LDAP repository at hand, a very light solution for testing is the `acme-ldap` java server included in the GeoServer LDAP documentation.

Reviewer: ```suggestion
If you don't have an LDAP repository available, a very light solution for testing is the `acme-ldap` java server included in the GeoServer LDAP documentation.
```
